blazar/plugins/devices/device_plugin.py

```diff
@@ -0,0 +1,652 @@
+                if device_id in device_allocations.keys():
+                    device_allocations[device_id].append({
+                        k: v for k, v in reservation.items()
+                        if k != 'device_ids'})
+
+        return device_allocations
+
+    def allocation_candidates(self, values):
+        self._check_params(values)
+        
+        device_ids = self._matching_devices(
+            values['device_properties'],
+            values['resource_properties'],
+            values['count_range'],
+            values['start_date'],
+            values['end_date']
+        )
+
+        min_devices, _ = [int(n) for n in values['count_range'].split('-')]
+
+        if len(device_ids) < min_devices:
+            raise manager_ex.NotEnoughHostsAvailable()
+
+        return device_ids
+    
+    def _convert_int_param(self, param, name):
+        """Checks that the parameter is present and can be converted to int."""
+        if param is None:
+            raise manager_ex.MissingParameter(param=name)
+        if strutils.is_int_like(param):
+            param = int(param)
+        else:
+            raise manager_ex.MalformedParameter(param=name)
+        return param
+    
+    def _validate_min_max_range(self, values, min_devices, max_devices):
+        min_devices = self._convert_int_param(min_devices, 'min')
+        max_devices = self._convert_int_param(max_devices, 'max')
+        if min_devices <= 0 or max_devices <= 0:
+            raise manager_ex.MalformedParameter(
+                param='min and max (must be greater than or equal to 1)')
+        if max_devices < min_devices:
+            raise manager_ex.InvalidRange()
+        values['count_range'] = str(min_devices) + '-' + str(max_devices)
+    
+    def _check_params(self, values):
+        self._validate_min_max_range(values, values.get('min'),
+                                     values.get('max'))
+
+        if 'device_properties' not in values:
+            raise manager_ex.MissingParameter(param='device_properties')
+        if 'resource_properties' not in values:
+            raise manager_ex.MissingParameter(param='resource_properties')
+
+        if 'before_end' not in values:
+            values['before_end'] = 'default'
+        if values['before_end'] not in before_end_options:
+            raise manager_ex.MalformedParameter(param='before_end')
+
+        if 'on_start' not in values:
+            values['on_start'] = 'default'
+
+    def _matching_devices(self, device_properties, resource_properties, 
+                          count_range, start_date, end_date):
+        """Return the matching devices (preferably not allocated)"""
+        count_range = count_range.split('-')
+        min_device = count_range[0]
+        max_device = count_range[1]
+        allocated_device_ids = []
+        not_allocated_device_ids = []
+        filter_array = []
+        start_date_with_margin = start_date - datetime.timedelta(
```

Comment:
Might want to make a separate config for cleaning_time, as cleaning time for nodes is higher than what we should need for containers/vm/shell.